packages/react-dom/src/client/ReactDOMRoot.js

```diff
@@ -115,6 +116,13 @@ ReactDOMRoot.prototype.unmount = ReactDOMBlockingRoot.prototype.unmount = functi
   });
 };
 
+ReactDOMRoot.prototype.registerMutableSourceForHydration = ReactDOMBlockingRoot.prototype.registerMutableSourceForHydration = function(
+  mutableSource: MutableSource<any>,
```

Comment:
Have you considered making his an option to `createRoot`? Like `hydrate` is.

Reply:
I hadn't considered it, no. Seems reasonable!

Reply:
Though I wonder if it's ever the case that you need to create the root before you know what all the mutable sources are. Like if there's a source that's only used by a chunk of the app and that chunk doesn't get streamed in until later. Seems plausible because otherwise you have to know about all possible sources up front, which is bad for code splitting.

Reply:
That's possible. Feels kind of odd to pass around the root on the other hand.

I don't have a strong preference between these two so I'll just leave it as is for the time being. Seb can share his thoughts.

Reply:
The `root.render(...)` API can be used for updates. The root instance is generally an "alive" thing once created.

It wouldn't make sense to call `registerMutableSourceForHydration` after hydration has already started since bets are off then. It would have to start erroring based on temporal values which seems a bit like a foot gun when you refactor or change the timing of two async calls like bootloading the code that runs render and the core that registerMutableSourceForHydration in parallel causing a race condition.

For event replaying I was thinking that createRoot would be passed a list of events that can be replayed after they've been recorded before React loaded. These are "grandfathered" into the event replaying queue. Once createRoot is called, we've already attached the event listeners on the root so it's a good time to hand off.

Additionally I was thinking createRoot would be passed a "time" option to indicate what the time was at the server render to get consistent hydration of time based values. It also already takes callbacks for hydration purposes too.

I.e. there's already a grab bag of option that go into createRoot and that's really the point of handoff from prehydration work to hydration. So I think this should also be an option to createRoot.

Reply:
> It wouldn't make sense to call `registerMutableSourceForHydration` after hydration has already started since bets are off then.

This is a strong reason. 👍 I'm convinced.